### ISSUE TITLE
Extract and make SQL commands a framework feature

### ISSUE THREAD

**Issue by electricjones** [ravenfire/magpie#18]
This is a follow-up to #7.

We have good examples of SQL commands. We should now move those SQLCommands to the `Ravenfire` portion and integrate the "idea" of a sql command into the framework itself.

The idea is that the user can define any number of sql commands in the `PrimaryEntity` or maybe `Source` classes and those commands will automatically be added to the cli under a group.

**Comment by JamesRichards07** [comment]
I'm guessing this issue is to allow Magpie users to add their own SQL commands. Do you have any resources I should look into before starting this issue?

**Comment by electricjones** [comment]
So, as I reread my ticket, I'm thinking 2 things:
1. Make it easy for users to add `Command` to their sources. Not just for sql. So, they can add any command. This would be pretty simple. You'd just define another method on either/both of the `PrimaryEntity` and/or `Source` that they can add their own classes to. Then, when we add the commands to the application, look through all of the "registered" ones for all the primary entity or sources and add those too.

2. Then, thinking about sql specifically, if you wanted, you could just extract some of your sql stuff into traits to make it easier to reuse.

Then, if someone wants to add their own "purge old records" command, they can. But if they want their own sql command, they just add a new command, include a trait or two and get some of the work for free.

**Comment by electricjones** [comment]
This is/can be connected with the "don't use so many static methods" ticket (#10).

This is what I mean.

If we use a static method for these commands (like we do for everything else) then its harder for them to setup their commands if they need to do anything extra like pass in something to the constructor. They don't actually have access to the code at that point.

But, if there were an instance method like `getExtraCommand(): array` that returns an array with some command objects all ready to go on the Source and PrimaryEntity, then we just cycle through the sources and primary entity and add those commands.

I can be more specific with my ideas in code. It's kinda hard to explain.

**Comment by JamesRichards07** [comment]
Yeah, either code or a Zoom would help. I'm trying to visualize what you mean but not sure I'm going in the right direction.

**Comment by JamesRichards07** [comment]
When you are talking about moving some of my sql stuff into traits I think you mean just pieces that are reused in every sql script like when you create a new table with rows, headers, and render()? 

I thought about creating specific methods for adding and get arguments and options but decided against it because those are pretty much already built in to Symphony.

**Comment by electricjones** [comment]
Yeah, I would agree with that. Just the simple boilerplate stuff. Stuff that you wouldn't want to figure out how to write again, and that symfony/magpie doesn't already give you.
